### PR TITLE
[fix][broker] Avoid compaction task stuck when the last message to compact is a marker

### PR DESCRIPTION
**Pull request by coderzc**
### Motivation

If the last message to compact is a marker, then the compaction task will stuck until timeout fails. The root cause is that the marker has been filtered out by the dispatcher, and the compaction reader will not read the last message.

```java
 WARN  - [broker-client-shared-internal-executor-18-1:PersistentTopic@3432] - [persistent://tnx/ns-prechecks/test_transaction_topic211ca13d-f8a4-4648-a7b9-bd2f679e0f3b] Compaction failure.
java.util.concurrent.CompletionException: org.apache.pulsar.common.util.FutureUtil$LowOverheadTimeoutException: Timeout
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:708) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162) ~[?:?]
	at org.apache.pulsar.common.util.FutureUtil.lambda$addTimeoutHandling$9(FutureUtil.java:236) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.100.Final.jar:4.1.100.Final]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: org.apache.pulsar.common.util.FutureUtil$LowOverheadTimeoutException: Timeout
	at org.apache.pulsar.compaction.TwoPhaseCompactor.phaseOneLoop(...)(Unknown Source) ~[classes/:?]
```

### Modifications

For `__compaction` internal cursor, we don't need to filter out marker messages and delete marker messages during compaction.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java [hunk 1]
```diff
@@ -1849,4 +1850,35 @@ public void testReadCommittedWithReadCompacted() throws Exception{
...
+        producer.newMessage(txn).key("K3").value("V3").send();
+        txn.commit();
+
+        admin.topics().triggerCompaction(topic);
+
+        Awaitility.await().untilAsserted(() -> {
+            assertEquals(admin.topics().compactionStatus(topic).status,
+                    LongRunningProcessStatus.Status.SUCCESS);
+        });
+    }
```
Comment (Technoboy-) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java:1882]: The current test case is : `normal` msg and then `txn`msg.    Could you also add `txn` msg and then `normal` msg ?

Comment (coderzc) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java:1909]: Done.

**Comment by Technoboy-** [comment]
We need to cherry to 2.9 ~ 3.1 right ?

**Comment by coderzc** [comment]
> We need to cherry to 2.9 ~ 3.1 right ?

Yes.

**Review by Technoboy- — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java [hunk 2]
```diff
@@ -192,7 +195,7 @@ public int filterEntriesForConsumer(@Nullable MessageMetadata[] metadataArray, i
                 }
             }
 
-            if (msgMetadata == null || Markers.isServerOnlyMarker(msgMetadata)) {
+            if (msgMetadata == null || (Markers.isServerOnlyMarker(msgMetadata) && !Markers.isTxnMarker(msgMetadata))) {
```
Comment (liangyepianzhou) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java:198]: I think we should pass all the markers and filter them at the client side when doing compaction.

File: pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java [hunk 3]
```diff
@@ -262,7 +266,10 @@ private void phaseTwoLoop(RawReader reader, MessageId to, Map<String, MessageId>
                 MessageId id = m.getMessageId();
                 Optional<RawMessage> messageToAdd = Optional.empty();
                 mxBean.addCompactionReadOp(reader.getTopic(), m.getHeadersAndPayload().readableBytes());
-                if (RawBatchConverter.isReadableBatch(m)) {
+                MessageMetadata metadata = Commands.parseMessageMetadata(m.getHeadersAndPayload());
+                if (Markers.isTxnMarker(metadata)) {
```
Comment (Demogorgon314) [inline pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java:270]: Why only check the txn marker here? Should we use `Markers.isServerOnlyMarker(metadata)`?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java [hunk 4]
```diff
@@ -1849,4 +1850,62 @@ public void testReadCommittedWithReadCompacted() throws Exception{
         Assert.assertEquals(messages, List.of("V2", "V3"));
     }
 
+
+    @Test
+    public void testReadCommittedWithCompaction() throws Exception{
```
Comment (Demogorgon314) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java:1855]: Should we add a test to send a replicated subscription snapshot marker and test the compaction works?

File: pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java (hunk 3 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java:270]: Yes, I will fix it.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java (hunk 4 above)
Comment (coderzc) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java:1855]: I add test `testReplicatedSubscriptionWithCompaction`, PLAT~

**Review by liangyepianzhou — APPROVED** [review]

**Review by Demogorgon314 — APPROVED** [review]

**Pull request merged**